gensim/corpora/textcorpus.py

```diff
@@ -98,28 +98,54 @@ def get_texts(self):
                 else:
                     yield utils.tokenize(line, lowercase=True)
 
-    def sample_texts(self, n):
+    def sample_texts(self, n, seed=None, length=None):
         """
-        Yield n random texts from the corpus without replacement.
+        Yields n random documents from the corpus without replacement.
 
-        Given the the number of remaingin elements in stream is remaining and we need
-        to choose n elements, the probability for current element to be chosen is n/remaining.
+        Given the number of remaining documents in corpus, we need to choose n elements.
+        The probability for current element to be chosen is n/remaining.
         If we choose it, we just decreese the n and move to the next element.
+        Computing corpus length may be a costly operation so you can use optional paramter
+        length instead.
+
+        Args:
+            n (int): number of documents we want to sample.
+            seed (int|None): if specified, use it as a seed for local random generator.
+            length (int|None): if specified, use it as guess of corpus length.
+
+        Yeilds:
+            list[str]: document represented as list of tokens. See get_texts method.
+
+        Raises:
+            ValueError: then n is invalid or length was set incorrectly.
         """
-        length = len(self)
-        if not n <= length:
-            raise ValueError("sample larger than population")
+        random_generator = None
+        if seed is None:
+            random_generator = random
+        else:
+            random_generator = random.Random(seed)
+
+        if length is None:
+            length = len(self)
 
+        if not n <= length:
+            raise ValueError("n is larger than length of corpus.")
         if not 0 <= n:
-            raise ValueError("negative sample size")
+            raise ValueError("Negative sample size.")
 
         for i, sample in enumerate(self.get_texts()):
-            remaining_in_stream = length - i
-            chance = random.randint(1, remaining_in_stream)
+            if i == length:
+                break
+            remaining_in_corpus = length - i
+            chance = random_generator.randint(1, remaining_in_corpus)
             if chance <= n:
                 n -= 1
                 yield sample
 
+        if n != 0:
+            # This means that length was set to be smaller than nuber of items in stream.
+            raise ValueError("length smaller than number of documents in stream")
+
```

Comment:
It should be _length was set to be **greater** than number of items in stream_  indeed.  

It throws an exception because user received less than `n` docs. Note that if `length` is greater than actual length of the corpus, it will not trigger the exception on its own. The exception triggers when we still need to sample some documents, but the stream run out.

The case when `length` was smaller than the actual corpus length does not throw anything right now (the problem was only in the documentation). Printing a warning is awkward, because we would have to actually go through all of the `length` elements which is unnecessary (we may sample enough elements earlier).